lib/boba/enums_railtie.rb

```diff
@@ -0,0 +1,100 @@
+# typed: true
+# frozen_string_literal: true
+
+require "rails/railtie"
+
+module Boba
+  class EnumsRailtie < Rails::Railtie
+    railtie_name(:boba)
+
+    initializer("boba.add_enum_classes") do
+      ActiveSupport.on_load(:active_record) do
+        next if defined?(ActiveRecordTypedEnum)
+
+        module ActiveRecordTypedEnum
+          extend T::Sig
+          extend T::Helpers
+
+          module ClassMethods
+            extend T::Sig
+            extend T::Helpers
+            include Kernel
+
+            abstract!
+
+            sig { abstract.returns(T::Hash[String, T::Hash[String, T.untyped]]) }
+            def defined_enums; end
+
+            sig { abstract.params(name: String, value: T.untyped).returns(T.untyped) }
+            def const_set(name, value)
+            end
+
+            sig { abstract.params(name: String, block: T.proc.params(args: T.untyped).void).returns(T.untyped) }
+            def define_method(name, &block)
+            end
+
+            sig { params(args: T.untyped, kwargs: T.untyped).returns(T.untyped) }
+            def enum(*args, **kwargs)
+              # Call the original enum method first
+              result = super
+
+              # Extract definitions based on how enum was called
+              definitions = if args.first.is_a?(Hash)
+                # Rails 6 style: enum(status: {draft: 0, published: 1})
+                args.first
+              elsif args.length >= 2
+                # Rails 7 style: enum(:status, {draft: 0, published: 1})
+                { args[0] => args[1] }
+              elsif kwargs.any?
+                # Keyword style: enum(status: {draft: 0, published: 1})
+                kwargs
+              else
+                {}
+              end
+
+              # Add typed enum methods for each enum
+              definitions.each_key do |name|
+                enum_values = defined_enums.fetch(name.to_s)
+
+                enum_class = Class.new(T::Enum) do
+                  enums { enum_values.each { |key, value| const_set(key.to_s.camelize, new(value)) } }
+                end
+                enum_class.define_method(:rails_enum_name) do
+                  T.bind(self, T::Enum)
+                  T.must(enum_values.find { |_, value| serialize == value }).first
+                end
+
+                const_set(name.to_s.camelize, enum_class)
+
+                # Define typed getter method
+                define_method("typed_#{name}") do
+                  value = send(name)
+                  return if value.nil?
+
+                  enum_variant = T.must(enum_values.find { |key, _| key == value })
```

Comment:
This is going the 'opposite' way as what I described above!

When we want to return the right enum variant, we get the Rails name from its default getter. And we need to find the right T::Enum serialized value, that we can pass the `try_deserialize` below